### PR TITLE
Allow parallelized feature extraction using joblib

### PR DESCRIPTION
**Pull request by hofaflo**

### CODE REVIEW

File: sleepecg/utils.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    function: Callable[..., _Returnable],
+    iterable: Iterable,
+    *args,
+    **kwargs,
+) -> List[_Returnable]:
+    """
+    Apply a function to each element in an iterable in parallel.
+
+    joblib is used for parallelism. If it's not available, this falls back
+    to a pure python loop.
```
Comment (cbrnr) [inline sleepecg/utils.py:26]:
```suggestion
    This uses joblib for parallelism. If the package is not available, it
    falls back to a pure Python loop.
```

File: sleepecg/utils.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    except ImportError:
+        warnings.warn('joblib not installed, cannot run in parallel.')
+        n_jobs = 1
+
+    if n_jobs == 1:
+        return [function(x, *args, **kwargs) for x in iterable]
+
+    return Parallel(n_jobs=n_jobs)(
+        delayed(function)(x, *args, **kwargs) for x in iterable
+    )
```
Comment (cbrnr) [inline sleepecg/utils.py:77]:
```suggestion
        return [function(x, *args, **kwargs) for x in iterable]

    return Parallel(n_jobs=n_jobs)(
        delayed(function)(x, *args, **kwargs) for x in iterable
    )
```

File: sleepecg/feature_extraction.py [hunk 3]
```diff
@@ -436,6 +437,132 @@ def preprocess_rri(
...
+        time in seconds.
+    lookforward : int, optional
+        Forward extension of the analysis window from each sleep stage
+        time in seconds.
+    fs_rri_resample : float
+        Frequency in Hz at which the RRI time series should be resampled
+        before spectral analysis. Only relevant for frequency domain
+        features.
+    max_nans : float
+        Maximum fraction of NaNs in an analysis window, for which frequency
```
Comment (cbrnr) [inline sleepecg/feature_extraction.py:485]:
```suggestion
        Maximum fraction of NaNs in an analysis window for which frequency
```

File: sleepecg/feature_extraction.py [hunk 4]
```diff
@@ -436,6 +437,132 @@ def preprocess_rri(
...
+        all `required_groups`. Required to select the columns corresponding
+        to the requested `feature_ids` from the calculated feature matrix.
+
+    Returns
+    -------
+    features : np.ndarray
+        The feature matrix of shape `(len(sleep_stages), <num_features>)`
+        containing the extracted features.
+    stages : np.ndarray | None
+        Tge label vector, i.e. the annotated sleep stages. For a `record`
```
Comment (cbrnr) [inline sleepecg/feature_extraction.py:503]:
```suggestion
        The label vector, i.e. the annotated sleep stages. For a `record`
```

File: sleepecg/feature_extraction.py [hunk 5]
```diff
@@ -436,6 +437,132 @@ def preprocess_rri(
...
+        without annotated stages, this will be `None`.
+    """
+    rri_required = 'hrv-time' in required_groups or 'hrv-frequency' in required_groups
+
+    if record.sleep_stages is not None and record.sleep_stage_duration is not None:
+        record_duration = (len(record.sleep_stages) - 1) * record.sleep_stage_duration
+    elif record.heartbeat_times is not None:
+        record_duration = record.heartbeat_times[-1]
+    else:
+        raise ValueError(f"Record duration can't be inferred for {record.id}. ")
```
Comment (cbrnr) [inline sleepecg/feature_extraction.py:513]:
```suggestion
        raise ValueError(f"Record duration cannot be inferred for {record.id}. ")
```

File: sleepecg/feature_extraction.py [hunk 6]
```diff
@@ -436,6 +437,132 @@ def preprocess_rri(
...
+    num_stages = record_duration // sleep_stage_duration
+    stage_times = np.arange(num_stages) * sleep_stage_duration
+
+    if rri_required:
+        if record.heartbeat_times is None:
+            raise ValueError(f"Can't extract HRV features for record {record.id} without heartbeat_times.")  # noqa: E501
```
Comment (cbrnr) [inline sleepecg/feature_extraction.py:519]:
```suggestion
            raise ValueError(f"Cannot extract HRV features for record {record.id} without heartbeat_times.")  # noqa: E501
```

File: sleepecg/feature_extraction.py [hunk 7]
```diff
@@ -522,65 +653,25 @@ def extract_features(
...
+        required_groups,
+        lookback,
+        lookforward,
+        fs_rri_resample,
+        max_nans,
+        feature_ids,
+        col_indices,
+    )
+    features = [X for X, _ in Xy]
+    stages = [y for _, y in Xy]
```
Comment (cbrnr) [inline sleepecg/feature_extraction.py:675]: Would that work too?

```suggestion
    features, stages = zip(*Xy)
```

Comment (hofaflo) [reply sleepecg/feature_extraction.py:675]: yes, but the types wouldn't propagate correctly :/

Comment (cbrnr) [reply sleepecg/feature_extraction.py:675]: Did I mention I really don't like type annotations in Python?

Comment (hofaflo) [reply sleepecg/feature_extraction.py:675]: once or twice :D

**Comment by cbrnr** [comment]
Thanks @hofaflo!

**Pull request merged**